File: python/cmd/dataset_ingester.py
#!/usr/bin/env python3
"""Deterministic dataset ingester for the Python journal writer."""

from __future__ import annotations

import argparse
import base64
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "python"))

from journal import Writer  # noqa: E402

BOOT_ID = bytes.fromhex("0123456789abcdef0123456789abcdef")
MACHINE_ID = bytes.fromhex("fedcba9876543210fedcba9876543210")
SEQNUM_ID = bytes.fromhex("22222222222222222222222222222222")
FILE_ID = bytes.fromhex("33333333333333333333333333333333")
OVERSIZED_LIMIT = 4 * 1024 * 1024
DEFAULT_ARCHIVE_REALTIME = 1_700_000_000_000_000


def materialize_value(value: dict) -> bytes:
    kind = value.get("kind")
    if kind == "utf8":
        return value["text"].encode("utf-8")
    if kind == "bytes":
        data = base64.b64decode(value["base64"])
        expected = value.get("size")
        if expected is not None and len(data) != expected:
            raise ValueError(f"bytes size mismatch: expected {expected}, got {len(data)}")
        return data
    if kind == "repeat":
        return bytes([value["byte"]]) * value["size"]
    raise ValueError(f"unknown value kind: {kind!r}")


def expected_rejection(input_data: dict) -> str | None:
    if "raw_payload" in input_data:
        raw = input_data["raw_payload"]
        if "=" not in raw:
            return "EINVAL"
        name, _value = raw.split("=", 1)
        if not valid_field_name(name):
            return "EINVAL"
        return None

    name = input_data.get("field_name")
    if name is None or not valid_field_name(name):
        return "EINVAL"
    value = input_data.get("value")
    if value is None:
        return "EINVAL"
    if isinstance(value, dict) and value.get("kind") == "repeat" and value.get("size", 0) > OVERSIZED_LIMIT:
        return "E2BIG"
    return None


def valid_field_name(name: str) -> bool:
    if not name or len(name.encode("utf-8")) > 64:
        return False
    first = name[0]
    if "0" <= first <= "9":
        return False
    return all(ch == "_" or "A" <= ch <= "Z" or "0" <= ch <= "9" for ch in name)


def make_writer(path: Path, compact: bool, max_size_bytes: int | None) -> Writer:
    path.parent.mkdir(parents=True, exist_ok=True)
    options = {
        "boot_id": BOOT_ID,
        "machine_id": MACHINE_ID,
        "seqnum_id": SEQNUM_ID,
        "file_id": FILE_ID,
        "head_seqnum": 1,
        "compression": "none",
        "compression_threshold_bytes": 512,
        "compact": compact,
    }
    if max_size_bytes is not None:
        options["max_file_size"] = max_size_bytes
    return Writer.create(str(path), options)


def archive_path_for(output: Path, head_realtime: int) -> Path:
    prefix = output.name[:-len(".journal")] if output.name.endswith(".journal") else output.name
    name = f"{prefix}@{SEQNUM_ID.hex()}-0000000000000001-{head_realtime:016x}.journal"
    return output.with_name(name)


def finalize_writer(writer: Writer, output: Path, final_state: str, head_realtime: int) -> None:
    if final_state == "online":
        writer.close()
    elif final_state == "offline":
        writer.close_offline()
    elif final_state == "archived":
        archive_path = archive_path_for(output, head_realtime)
        archive_path.unlink(missing_ok=True)
        writer.archive_to(str(archive_path))
    else:
        raise ValueError(f"invalid final state: {final_state}")


def ingest_accepted(
    dataset: Path,
    output: Path,
    final_state: str,
    compact: bool,
    max_size_bytes: int | None,
) -> dict:
    writer = make_writer(output, compact, max_size_bytes)
    written = 0
    head_realtime = 0
    errors: list[str] = []
    try:
        with dataset.open("r", encoding="utf-8") as f:
            for line_no, line in enumerate(f, 1):
                if not line.strip():
                    continue
                record = json.loads(line)
                if record.get("record_type") != "accepted":
                    continue
                fields = [
                    {"name": item["name"], "value": materialize_value(item["value"])}
                    for item in record["fields"]
                ]
                try:
                    writer.append(
                        fields,
                        {
                            "realtime_usec": record["realtime_usec"],
                            "monotonic_usec": record["monotonic_usec"],
                            "boot_id": bytes.fromhex(record.get("boot_id", BOOT_ID.hex())),
                        },
                    )
                    if head_realtime == 0:
                        head_realtime = record["realtime_usec"]
                    written += 1
                except Exception as err:  # pragma: no cover - command line diagnostic
                    errors.append(f"line {line_no}: append failed: {err}")
        writer.sync()
    finally:
        finalize_writer(writer, output, final_state, head_realtime or DEFAULT_ARCHIVE_REALTIME)

    return {"records": written, "errors": errors}


def ingest_rejections(
    dataset: Path,
    output: Path,
    final_state: str,
    compact: bool,
    max_size_bytes: int | None,
) -> dict:
    writer: Writer | None = None
    handled = 0
    errors: list[str] = []

    with dataset.open("r", encoding="utf-8") as f:
        for line_no, line in enumerate(f, 1):
            if not line.strip():
                continue
            record = json.loads(line)
            if record.get("record_type") != "rejected":
                continue
            case_id = record["case_id"]
            expected = record["expected_error"]
            input_data = record.get("input", {})
            precheck = expected_rejection(input_data)
            if precheck is not None:
                if precheck == expected:
                    handled += 1
                else:
                    errors.append(f"line {line_no} {case_id}: got {precheck}, expected {expected}")
                continue

            if writer is None:
                writer = make_writer(output, compact, max_size_bytes)
            name = input_data["field_name"]
            value = materialize_value(input_data["value"])
            try:
                writer.append([{"name": name, "value": value}], {"boot_id": BOOT_ID})
                errors.append(f"line {line_no} {case_id}: unexpectedly accepted")
            except Exception:
                if expected == "EINVAL":
                    handled += 1
                else:
                    errors.append(f"line {line_no} {case_id}: rejected as EINVAL, expected {expected}")

    if writer is not None:
        finalize_writer(writer, output, final_state, DEFAULT_ARCHIVE_REALTIME)
    return {"records": handled, "errors": errors}


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--dataset", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--rejection-mode", action="store_true")
    parser.add_argument("--final-state", choices=("online", "offline", "archived"), default="online")
    parser.add_argument("--compact", action="store_true")
    parser.add_argument("--max-size-bytes", type=int)
    args = parser.parse_args()

    result = (
        ingest_rejections(args.dataset, args.output, args.final_state, args.compact, args.max_size_bytes)
        if args.rejection_mode
        else ingest_accepted(args.dataset, args.output, args.final_state, args.compact, args.max_size_bytes)
    )
    print(json.dumps(result, sort_keys=True))
    return 0 if not result["errors"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: experiments/node/cmd/dataset_ingester.js
#!/usr/bin/env node
// Deterministic dataset ingester for the JavaScript journal writer.

import { readFileSync, rmSync } from 'node:fs';
import { mkdirSync } from 'node:fs';
import { dirname } from 'node:path';
import { Writer } from '../src/lib/writer.js';

const BOOT_ID = Buffer.from('0123456789abcdef0123456789abcdef', 'hex');
const MACHINE_ID = Buffer.from('fedcba9876543210fedcba9876543210', 'hex');
const SEQNUM_ID = Buffer.from('22222222222222222222222222222222', 'hex');
const FILE_ID = Buffer.from('33333333333333333333333333333333', 'hex');
const OVERSIZED_LIMIT = 4 * 1024 * 1024;
const DEFAULT_ARCHIVE_REALTIME = 1_700_000_000_000_000n;

function materializeValue(value) {
  if (value.kind === 'utf8') return Buffer.from(value.text, 'utf8');
  if (value.kind === 'bytes') {
    const data = Buffer.from(value.base64, 'base64');
    if (value.size !== undefined && data.length !== value.size) {
      throw new Error(`bytes size mismatch: expected ${value.size}, got ${data.length}`);
    }
    return data;
  }
  if (value.kind === 'repeat') return Buffer.alloc(value.size, value.byte);
  throw new Error(`unknown value kind: ${value.kind}`);
}

function validFieldName(name) {
  if (!name || Buffer.byteLength(name, 'utf8') > 64) return false;
  const first = name.charCodeAt(0);
  if (first >= 0x30 && first <= 0x39) return false;
  for (let i = 0; i < name.length; i++) {
    const ch = name.charCodeAt(i);
    if (ch !== 0x5f && !(ch >= 0x41 && ch <= 0x5a) && !(ch >= 0x30 && ch <= 0x39)) {
      return false;
    }
  }
  return true;
}

function expectedRejection(input) {
  if (input.raw_payload !== undefined) {
    const eq = input.raw_payload.indexOf('=');
    if (eq < 0) return 'EINVAL';
    const name = input.raw_payload.slice(0, eq);
    return validFieldName(name) ? null : 'EINVAL';
  }
  const name = input.field_name;
  if (name === undefined || !validFieldName(name)) return 'EINVAL';
  const value = input.value;
  if (value === null || value === undefined) return 'EINVAL';
  if (value.kind === 'repeat' && value.size > OVERSIZED_LIMIT) return 'E2BIG';
  return null;
}

function makeWriter(path, compact, maxSizeBytes) {
  // eslint-disable-next-line security/detect-non-literal-fs-filename -- output path is the explicit CLI target.
  mkdirSync(dirname(path), { recursive: true });
  const options = {
    bootId: BOOT_ID,
    machineId: MACHINE_ID,
    seqnumId: SEQNUM_ID,
    fileId: FILE_ID,
    headSeqnum: 1,
    compression: 'none',
    compressionThresholdBytes: 512,
    compact,
  };
  if (maxSizeBytes !== undefined) options.maxFileSize = maxSizeBytes;
  return Writer.create(path, options);
}

function archivePathFor(output, headRealtime) {
  const prefix = output.endsWith('.journal') ? output.slice(0, -'.journal'.length) : output;
  return `${prefix}@${SEQNUM_ID.toString('hex')}-0000000000000001-${headRealtime.toString(16).padStart(16, '0')}.journal`;
}

function finalizeWriter(writer, output, finalState, headRealtime) {
  if (finalState === 'online') writer.close();
  else if (finalState === 'offline') writer.closeOffline();
  else if (finalState === 'archived') {
    const archivePath = archivePathFor(output, headRealtime);
    rmSync(archivePath, { force: true });
    writer.archiveTo(archivePath);
  }
  else throw new Error(`invalid final state: ${finalState}`);
}

function records(path) {
  // eslint-disable-next-line security/detect-non-literal-fs-filename -- dataset path is the explicit CLI input.
  return readFileSync(path, 'utf8')
    .split('\n')
    .filter(Boolean)
    .map(line => JSON.parse(line));
}

function ingestAccepted(dataset, output, finalState, compact, maxSizeBytes) {
  const writer = makeWriter(output, compact, maxSizeBytes);
  let written = 0;
  let headRealtime = 0n;
  const errors = [];
  try {
    for (const [index, record] of records(dataset).entries()) {
      if (record.record_type !== 'accepted') continue;
      const fields = record.fields.map(item => ({
        name: item.name,
        value: materializeValue(item.value),
      }));
      try {
        writer.append(fields, {
          realtimeUsec: BigInt(record.realtime_usec),
          monotonicUsec: BigInt(record.monotonic_usec),
          bootId: Buffer.from(record.boot_id || BOOT_ID.toString('hex'), 'hex'),
        });
        if (headRealtime === 0n) headRealtime = BigInt(record.realtime_usec);
        written++;
      } catch (error) {
        errors.push(`line ${index + 1}: append failed: ${error.message}`);
      }
    }
    writer.sync();
  } finally {
    finalizeWriter(writer, output, finalState, headRealtime || DEFAULT_ARCHIVE_REALTIME);
  }
  return { records: written, errors };
}

function ingestRejections(dataset, output, finalState, compact, maxSizeBytes) {
  let writer = null;
  let handled = 0;
  const errors = [];
  for (const [index, record] of records(dataset).entries()) {
    if (record.record_type !== 'rejected') continue;
    const expected = record.expected_error;
    const precheck = expectedRejection(record.input || {});
    if (precheck !== null) {
      if (precheck === expected) handled++;
      else errors.push(`line ${index + 1} ${record.case_id}: got ${precheck}, expected ${expected}`);
      continue;
    }

    writer ||= makeWriter(output, compact, maxSizeBytes);
    try {
      writer.append(
        [{ name: record.input.field_name, value: materializeValue(record.input.value) }],
        { bootId: BOOT_ID },
      );
      errors.push(`line ${index + 1} ${record.case_id}: unexpectedly accepted`);
    } catch {
      if (expected === 'EINVAL') handled++;
      else errors.push(`line ${index + 1} ${record.case_id}: rejected as EINVAL, expected ${expected}`);
    }
  }
  if (writer) finalizeWriter(writer, output, finalState, DEFAULT_ARCHIVE_REALTIME);
  return { records: handled, errors };
}

function parseArgs(argv) {
  const args = Object.assign(Object.create(null), {
    finalState: 'online',
    rejectionMode: false,
    compact: false,
  });
  for (let i = 2; i < argv.length; i++) {
    const arg = argv.at(i);
    i = parseIngesterArg(args, argv, i, arg);
  }
  validateIngesterArgs(args);
  return args;
}

function parseIngesterArg(args, argv, index, arg) {
  if (arg === '--rejection-mode') { args.rejectionMode = true; return index; }
  if (arg === '--compact') { args.compact = true; return index; }
  const value = argv.at(index + 1);
  if (value === undefined) throw new Error(`missing value for ${arg}`);
  if (arg === '--final-state') args.finalState = value;
  else if (arg === '--dataset') args.dataset = value;
  else if (arg === '--output') args.output = value;
  else if (arg === '--max-size-bytes') args.maxSizeBytes = Number(value);
  else throw new Error(`unknown argument: ${arg}`);
  return index + 1;
}

function validateIngesterArgs(args) {
  if (!['online', 'offline', 'archived'].includes(args.finalState)) throw new Error(`invalid final state: ${args.finalState}`);
  if (args.maxSizeBytes !== undefined && (!Number.isSafeInteger(args.maxSizeBytes) || args.maxSizeBytes <= 0)) throw new Error('invalid --max-size-bytes');
  if (!args.dataset || !args.output) throw new Error('usage: dataset_ingester --dataset PATH --output PATH [--rejection-mode] [--final-state online|offline|archived] [--compact] [--max-size-bytes BYTES]');
}

try {
  const args = parseArgs(process.argv);
  const result = args.rejectionMode
    ? ingestRejections(args.dataset, args.output, args.finalState, args.compact, args.maxSizeBytes)
    : ingestAccepted(args.dataset, args.output, args.finalState, args.compact, args.maxSizeBytes);
  console.log(JSON.stringify(result, Object.keys(result).sort()));
  process.exit(result.errors.length === 0 ? 0 : 1);
} catch (error) {
  console.error(error.message);
  process.exit(1);
}
